File: LeetCode/205_Isomorphic_Strings/205_Isomorphic_Strings.kt
class Solution {
    fun isIsomorphic(s: String, t: String): Boolean {
        if (s.toCharArray().toSet().size != t.toCharArray().toSet().size) return false

        val hash: MutableMap<Char, Char> = mutableMapOf()
        for (i in 0 until s.length) {
            if (hash.contains(s[i]) && hash[s[i]] != t[i]) return false
            hash[s[i]] = t[i]
        }
        return true
    }
}

fun main() {
    val s = Solution()

    println(s.isIsomorphic("foo", "bar"))
    println(s.isIsomorphic("bacd", "baba"))
    println(s.isIsomorphic("paper", "title"))
}

File: LeetCode/205_Isomorphic_Strings/205_Isomorphic_Strings.go
package main

import "fmt"

type ns struct{}

func add(s map[byte]ns, v byte) map[byte]ns {
	s[v] = ns{}
	return s
}
func isIsomorphic(s string, t string) bool {
	set_s := make(map[byte]ns)
	set_t := make(map[byte]ns)
	for i := range s {
		set_s = add(set_s, s[i])
		set_t = add(set_t, t[i])
	}
	if len(set_s) != len(set_t) {
		return false
	}
	hash := make(map[byte]byte)
	for i := range s {
		val, ok := hash[s[i]]
		if ok && val != t[i] {
			return false
		}
		hash[s[i]] = t[i]
	}
	return true
}

func main() {

	fmt.Println(isIsomorphic("foo", "bar"))
	fmt.Println(isIsomorphic("bacd", "baba"))
	fmt.Println(isIsomorphic("paper", "title"))
}
